### DISCUSSION TITLE
Asymmetric draw aversion/inclination

### DISCUSSION THREAD

**Discussion by h8d13** [official-stockfish/Stockfish#6736] [Q&A]
Hi,

I've been lurking around chess automation for a while, mainly interested in draw theory: and recently read Qi (2026): arXiv:2604.03683 on asymmetric draw scoring.

As I was researching this, I came across issues #2953 and #746, where the comments themselves seem related: 

> "high draw rate is not only associated with strong play, it's also associated with safe play" / "When both engines draw an opening, the better engine draws in fewer moves (on defense)"

The paper's core idea is simple: (Altho it wants to put this at the rule level) if a Black draw is worth more than a White draw (since Black starts at a disadvantage), the scoring should reflect that  _somehow_. Perhaps it's something that has to be applied at the engine/testing level? 

If Black was improving faster at holding than White was at converting, the current scoring would be blind to that? Has anyone looked at the draw rate increase split by color around these periods of development/testing, the issue itself pointed to ` uc1 LTC sits at 86.3% draw rate. The completed uc3 LTC is at 86.5%.`

Looking at the LLR calculation in fishtest, draws are mapped to exactly `0.5` regardless of color (`LLRcalc.py` L227). 

Has anyone tried running patches with color-aware draw weights to see: if it changes which ones pass SPRT especially for patches that were rejected marginally ?

Discussion #4150 showed that the eval-result relationship is already asymmetric even in self-play, If the engine itself **doesn't** treat positions symmetrically, shouldn't results follow the same principle ? I also see that even the NNUE rep is also asym #6470

Then the benchmark would be how many draws did black pull-off and how many times could white convert its lead. So many patches that were increasing draws for black might have been missed ? It kind of makes sense, in my head at least.

I'm not sure how to test concretely aside from local builds from specific patches that had very high draw rates and digging here in the repo, what changed in between. Especially when code has to do with with draws.

Anyways, hopefully someone who knows more than I do can answer my question. Sorry if my thoughts seem a bit everywhere and thanks in advance.

**Comment by GuilhermeUcadete** [comment]
Interesting idea, but in practice this is likely a mismatch between game-theoretic intuition and SPRT-based testing goals. Fishtest isn’t trying to measure “which side exploits advantage better,” it’s trying to estimate overall Elo difference under symmetric play conditions. Assigning draws a fixed 0.5 enforces that neutrality.
While it’s true that draws aren’t perfectly symmetric in meaning (a Black draw can feel “better”), this asymmetry is already implicitly captured by Elo via win/loss rates. If a patch really improves Black’s defensive ability more than White’s conversion, it should still show up as a net Elo gain across many games.
Introducing color-dependent draw weights would:

break the clean statistical model behind SPRT,
complicate interpretation,
and risk biasing results toward specific playstyles rather than strength.

That said, your intuition is valuable: color-split metrics (draw rate, win rate, conversion rate) could be useful as secondary diagnostics, even if not used directly in SPRT decisions.

**Reply by h8d13** [reply]
Hi, thanks for the detailed answer. 

But per your own response: SPRT to my understanding uses

Win = 1 point, Draw = 0.5 points, Loss = 0 points. + replayed Pentanomial pairs `{0, 0.5, 1, 1.5, 2}`

This in essence lose 100% of the color signal (replayed with swap). It only asks “did the total score go up?” not “did you (with disadvantage) lose less/draw more?” 

As per mentioned issue #2953 The closest they got:

"high draw rate is not only associated with strong play, it's also associated with safe play" (incorrect IMO) then "When both engines draw an opening, the better engine draws in **fewer moves** (on defense.)"

I suspect pentanomial is the real issue in Fishtest: The pair system exists to eliminate color noise by averaging it out. I (and the paper) argues the color noise IS the signal. Fastchess (used by CPU contributors).

```shell

fastchess --help 
(truncated)
· -noswap
        Prevent swapping of colors.
```

This would allow for white/black fish specific patches?

Aside from this I believe the same is true in the https://github.com/official-stockfish/nnue-pytorch file: model/lightning_module.py:224 training of the nnue itself?

**Reply by GuilhermeUcadete** [reply]
Thanks for the thoughtful points. 😊
You’re right that pentanomial + color swapping is designed to remove color bias by construction, so SPRT ends up testing “net Elo improvement” rather than isolating performance under disadvantage. That’s intentional in Fishtest: the goal is to reduce variance and speed up convergence, even if it discards asymmetric signal.
That said, I agree your point is interesting, if color asymmetry actually carries useful information (e.g., robustness under pressure or defensive strength), then averaging it away could hide certain improvements. Using --noswap or asymmetric testing could surface that, but at the cost of much higher variance and potentially misleading results due to opening/color bias.
So it’s less that pentanomial is “wrong,” and more that it optimizes for statistical efficiency over expressiveness of the signal. Exploring separate white/black performance or conditional metrics could be a useful complementary approach, rather than replacing the current setup.
On the NNUE side, the same philosophy applies: the training pipeline aims for overall strength, not explicitly modeling color asymmetry. If that asymmetry matters, it could be an interesting direction for experimentation. but it’s not what the current framework is trying to optimize.

**Reply by h8d13** [reply]
Right so I'm not completely insane, the signal is lost end-to-end both in testing results, input and training.

On "could be useful as secondary diagnostics" I would argue that in "perfect play" as @nickolasreynolds suggested. it should be something to do with the amount of moves (momentum) and starting side.

Closing a draw from defense in sub 20 moves is not exactly the same as the typical 150 moves Engine v engine games.
Also, the data is already in huggingface PGN files:
```
[White "New-7649384e49"]   ← "New-" or "Base-" prefix
[Black "Base-39da50ed23"]
[Result "1-0"]             ← 1-0, 0-1, or 1/2-1/2
```
You can still today go find rejected/accepted that archived patches and parse with color signal.

Aside your point about opening books is valid: in a BlackFish vs WhiteFish world both should have their preferred evolving sets instead of Unbalanced Human Openings.

---

About you initial response "this asymmetry is already implicitly captured by Elo via win/loss rates." is I believe incorrect.  So The "asymmetry shows up in net Elo" only holds when the per-color shift is monotone in one direction.

On "could be useful as secondary diagnostics", that's exactly the proposal, I'm not asking to change SPRT decisions or the statistical model. Just report per-color pentanomial (noswap equivalents) alongside the pair vector.

**Reply by sscg13** [reply]
Why should we care about color? The point of a game pair is precisely to remove the dependence on color or bias. Chess itself is symmetric.

**Reply by h8d13** [reply]
Because white has a one move lead in every game out there: 

CCRL 3600elo+ reports (White WR, Draws, Black WR):  `7.9%   91.0%    1.1%`

So the asymmetry is already there, just not at the testing/input/training level. My understanding is that we plateau at such a high draw rate because we don't give NNUE signal of what drawing means per color. 

I disagree that chess is symmetric, this program, tests and NNUE are.

**Comment by vdbergh** [comment]
I don't understand why you think black is special. In the opening position white is slightly ahead and so a draw is indeed desirable for black. However in a position (e.g. a book exit) where black is better, a draw is desirable for white. So whether a draw is desirable or not is a function of the evaluation of the position and this is precisely what SF computes.

**Reply by h8d13** [reply]
I was not referring to black only: 

You're right that per-position eval already captures who's currently advantaged. The naive "all black draws +ε" framing double-counts when black is winning at the current position.

What I think survives: the training target with 0/0.5/1 outcome with no trajectory information.  

The narrative from the [paper](https://arxiv.org/pdf/2604.03683) and I want to support with this discussion: _"was this draw a 30-move forced repetition by the defending side, or a 150-move shuffled draw?"_  These for even for top-level human chess players are strictly different things. The former is a power move, the latter is just frustratingly long game-play that doesn't feel like a win (even if on paper for black it is).

For instance a black draw in sub 30 moves is a clear signal, just like a white draw 3-fold means it fumbled the bag (over move advantage). Per-color was my naive proxy for "did this side carry a temp deficit through the game"; move-count + eval-trajectory shape/over time is what I'm after.  + White 3-fold from the paper as clean penalty. 

As you said current pipeline labels both 0.5. UHO_+0.9 books also bias the test distribution toward white-favored starts, which compounds with the symmetric scoring (and would be interesting to test with noswaps + a black flavored book counterpart).

Testing methodology/results here also are directly tied, this is the part of the @nickolasreynolds observation in #2953 that I think has legs.

Proving it from the black-side, was my way to "challenge" the theory. Which made do some experiments here: https://github.com/h8d13/solid-engine/commit/26057afcfc8d3b90c8bb640629d7d9f7ebf6b52a  at the search level (last month).

That same day it drew 123 games in row against stockfish/reckless dev instances **as black only** (statistically improbable). (https://lichess.org/@/h8d13)

_"A draw is desirable for white"_ is not true in the context of this paper/perfect engine play: especially when through repetition, but also has no idea which kind/when draw happened. And the signal is lost in testing/training regardless.

If I'm understanding the full stack correctly: all 3 major components here assume symmetry (pair-averaging, boardflip in NNUE, training outcomes). 

I'm not an expert in these components (just cloned all the repos and searched for a month inside them) hence why I wanted to engage questions as to asymmetric "other side".
